src/VisualStudio/Core/Def/Implementation/UnusedReferences/ProjectAssets/ProjectAssetsReader.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis.UnusedReferences;
+using Newtonsoft.Json;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.UnusedReferences.ProjectAssets
+{
+    // This class will read the dependency heirarchy from the project.assets.json file. The format of this file
+    // is subject to change and in the future this information will be provided by an API with the ProjectSystem.
+    internal static class ProjectAssetsReader
+    {
+        public static ImmutableArray<ReferenceInfo> ReadReferences(
+            ImmutableArray<ReferenceInfo> projectReferences,
+            string projectAssetsFilePath,
+            string targetFrameworkMoniker)
+        {
+            var projectAssetsFileContents = File.ReadAllText(projectAssetsFilePath);
+            var projectAssets = JsonConvert.DeserializeObject<ProjectAssetsFile>(projectAssetsFileContents);
+
+            if (projectAssets is null ||
+                projectAssets.Version != 3)
+            {
+                return ImmutableArray<ReferenceInfo>.Empty;
+            }
+
+            if (projectAssets.Targets is null ||
+                !projectAssets.Targets.TryGetValue(targetFrameworkMoniker, out var target))
+            {
+                return ImmutableArray<ReferenceInfo>.Empty;
+            }
+
+            var autoReferences = projectAssets.Project?.Frameworks?.Values
+                .SelectMany(framework => framework.Dependencies?.Keys.Where(key => framework.Dependencies[key].AutoReferenced))
+                .Distinct()
+                .ToImmutableHashSet();
+            autoReferences ??= ImmutableHashSet<string>.Empty;
+
+            var references = projectReferences
+                .Select(projectReference => BuildReference(projectAssets, target, projectReference, autoReferences))
+                .OfType<ReferenceInfo>()
```

Comment:
Does `.OfType<>` do null checks? If not, this seems out of place.

Reply:
`OfType<>` does remove null items.

Reply:
It does

https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMA6AGQEsoBHAbjz1QGYT0iBhIgbzyIEk2qACxEAsgAoAlL36CF8gQF8lRNRnRq+uBQoD2AIwBWAUwDGwANoBdIkbOWAzkQC8RKKYDu9kxet2PEQARACCAA7hADamwTAhAPJgAIZQAOax8VAQUVFEyoy6egJqCgBuyWD2AGYAKgCe4aZwbr6OwE5kCXWNpgA8pAQAfDKFxQKkAJySg2QAUvo0knFxNQ1NcNLSY4KquMpAA===

Reply:
TIL, though it makes sense logically.